tools/check_documentation.py

```diff
@@ -0,0 +1,170 @@
+"""Checks for attributes and code examples in RedditBase subclasses"""
+
+import os
+import re
+import sys
+
+# This line imports from the local PRAW rather than the global installed PRAW.
+sys.path.insert(0, os.path.abspath(os.path.join(__file__, "..", "..")))
+
+from praw.util.cache import cachedproperty  # noqa: E402
+from praw.models.reddit.base import RedditBase  # noqa: E402
+from praw.models.reddit.modmail import ModmailObject  # noqa: E402
+
+
+class DocumentationChecker:
+    """Checks for code block statements and attribute tables in subclasses.
+
+    Attribute EXCEPTIONS holds all exceptions. All classes and subclasses in
+    the exceptions list will be ignored.
+
+    Attribute EXPLICIT_INCLUDE holds all classes that should be included
+    regardless of what is in the exceptions.
+
+    Attribute METHOD_EXCEPTIONS holds the names of methods that will be
+    filtered out.
+    """
+
+    HAS_CODE_BLOCK = re.compile(r".. code-block::")
+    HAS_ATTRIBUTE_TABLE = re.compile(r"Attribute[ ]+Description")
+
+    BASE_SEARCH_CLASS = RedditBase
+    SUBCLASSES = []
+    EXCEPTIONS = [
+        ModmailObject,  # is never publicly accessed
+    ]
+    EXPLICIT_INCLUDE = []
+
+    METHOD_EXCEPTIONS = [
+        "parse",
+        "sluggify",
+        "id_from_url",
+        "from_data",
+    ]
+
+    @staticmethod
+    def in_list_lambda(search_list):
+        return lambda item: item not in search_list
+
+    @classmethod
+    def expand_subclasses(cls):
+        """Recursively find subclasses"""
+        base_list = list(cls.BASE_SEARCH_CLASS.__subclasses__())
+        while len(base_list) > 0:
+            item = base_list[0]
+            base_list = base_list[1:]
+            cls.SUBCLASSES.append(item)
+            base_list.extend(item.__subclasses__())
+
+    @classmethod
+    def expand_exceptions(cls):
+        """Recursively find exceptions"""
+        base_list = cls.EXCEPTIONS
+        cls.EXCEPTIONS = []
+        while len(base_list) > 0:
+            item = base_list[0]
+            base_list = base_list[1:]
+            cls.EXCEPTIONS.append(item)
+            base_list.extend(item.__subclasses__())
+
+    @classmethod
+    def check(cls):
+        cls.expand_subclasses()
```

Comment:
These are run separately, because I want to include the methods of all subclasses. For example, `SubredditMessage` subclassess `Message`, but it won't show up without the expansion.

Reply:
Of course. That makes sense.